keps/sig-api-machinery/20190226-network-proxy.md

```diff
@@ -0,0 +1,577 @@
+    int64 connectionID = 1;
+
+    // Error when close request cannot be fulfilled.
+    Error error = 2;
+}
+
+message Data {
+    // connectID that data or error payload is belonged to
+    int32 connectID = 1;
+
+    // data payload
+    bytes data = 2;
+
+    // Error that happens to the connection.
+    Error error = 3;
+}
+```
+
+### Connectivity Proxy
+
+The connectivity proxy(s) can run in the same container as the KAS. 
+It should run on the same machine and must run in the same flat network as the KAS. 
+It listens on a port for gRPC connections from the KAS. 
+This port would be for forwarding traffic to the appropriate cluster. 
+It should have an administrative port speaking https. 
+The administrative port serves the liveness probe and metrics. 
+The liveness probe prevents a partially broken cluster 
+where the KAS cannot connect to the cluster. This port also serves 
+pprof debug commands and monitoring data for the proxy. 
+
+### Direct Connection
+
+This connection type uses the default dialer. 
+This allows use of the connectivity service without the connectivity proxy. 
+This is a quick way to run the system in a “legacy” or fallback mode. 
+Simple clusters (not needing network segregation) run this way to avoid the overhead 
+(in latency or configuration) of the connectivity proxy.
+
+### Kubernetes API Server Outbound Requests
+
+The majority of the KAS communication originates from incoming requests. 
+Here we cover the outgoing requests. This is our understanding of those requests 
+and some details as to how they fit in this model.
+
+- **ETCD** It is possible to make etcd talk via the proxy. 
+The etcd client takes a transport. 
+(https://github.com/etcd-io/etcd/blob/master/client/client.go#L101) 
+We will add configuration as to which proxy an etcd client should use. 
+(https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/storagebackend/config.go) 
+This will add an extra process hop to our main scaling axis. 
+We will scale test the impact and publish the results. As a precaution
+we will add an extra network configuration 'etcd' separate from ‘master’.
+Etcd requests can be configured separately from the rest of 'master'.  
+- **Pods/Exec**, **Pods/Proxy**, **Pods/Portforward**, **Pods/Attach**, **Pods/Log**
+Pod requests (and pod sub-resource requests) are meant for the cluster 
+and will be routed based on the ‘cluster’ NetworkContext.
+- **Nodes/Proxy**
+Node requests (and node sub-resource requests) are meant for the cluster 
+and will be routed based on the ‘cluster’ NetworkContext.
+- **Services/Proxy**
+Service requests (and service sub-resource requests) are meant for the cluster 
+and will be routed based on the ‘cluster’ NetworkContext.
+- **Admission Webhooks**
+Admission webhooks can be destined for a service or a URL. 
+If destined for a service then the service rules apply (send to 'cluster'). 
+If destined for a URL then we will use the ‘master’ NetworkContext. 
```

Comment:
Access to the master network via a URL on an admission webhook is an already existing behavior today, even when SSH Tunnnels have been configured.
Several features/customers rely on this behavior to access cloud provider specific webhooks running on some form of WebhookManager. 
So as it stands it is no worse than today and forwarding this traffic to the node network would break those features.
In addition the cloud providers can add filters to the master proxy to prevent access to things like etcd.
Lastly the traffic being sent is not arbitrary. The requests are admission requests formed by the API Server itself. We should not rely on being well formed admission requests but it does provide a last line of defense if something does get by the filter.